# IMPORT Libraries

In [5]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split

# READ CSV

In [6]:
my_data = pd.read_csv("Breast_Cancer.csv", delimiter=",")
my_data[0:5]

Age   Race Marital Status T Stage  N Stage 6th Stage  \
0   68  White        Married       T1      N1       IIA   
1   50  White        Married       T2      N2      IIIA   
2   58  White       Divorced       T3      N3      IIIC   
3   58  White        Married       T1      N1       IIA   
4   47  White        Married       T2      N1       IIB   

               differentiate Grade   A Stage  Tumor Size Estrogen Status  \
0      Poorly differentiated     3  Regional           4        Positive   
1  Moderately differentiated     2  Regional          35        Positive   
2  Moderately differentiated     2  Regional          63        Positive   
3      Poorly differentiated     3  Regional          18        Positive   
4      Poorly differentiated     3  Regional          41        Positive   

  Progesterone Status  Regional Node Examined  Reginol Node Positive  \
0            Positive                      24                      1   
1            Positive                      14                      5   
2            Positive                      14                      7   
3            Positive                       2                      1   
4            Positive                       3                      1   

   Survival Months Status  
0               60  Alive  
1               62  Alive  
2               75  Alive  
3               84  Alive  
4               50  Alive

# Gain some Information

In [7]:
my_data['Status'].unique()

array(['Alive', 'Dead'], dtype=object)

In [8]:
my_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4024 entries, 0 to 4023
Data columns (total 16 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Age                     4024 non-null   int64 
 1   Race                    4024 non-null   object
 2   Marital Status          4024 non-null   object
 3   T Stage                 4024 non-null   object
 4   N Stage                 4024 non-null   object
 5   6th Stage               4024 non-null   object
 6   differentiate           4024 non-null   object
 7   Grade                   4024 non-null   object
 8   A Stage                 4024 non-null   object
 9   Tumor Size              4024 non-null   int64 
 10  Estrogen Status         4024 non-null   object
 11  Progesterone Status     4024 non-null   object
 12  Regional Node Examined  4024 non-null   int64 
 13  Reginol Node Positive   4024 non-null   int64 
 14  Survival Months         4024 non-null   int64 
 15  Stat

# Create DataFrame

In [9]:
df=pd.DataFrame(my_data)

#### Information about quantity of values for each column

In [10]:
for col in df.columns:
    if not pd.api.types.is_integer_dtype(df[col]):  # Check if column is not integer type
        print(f"Value counts for {col}:")
        print(df[col].value_counts())
        print()  # Print an empty line for separation

Value counts for Race:
White    3413
Other     320
Black     291
Name: Race, dtype: int64

Value counts for Marital Status:
Married      2643
Single        615
Divorced      486
Widowed       235
Separated      45
Name: Marital Status, dtype: int64

Value counts for T Stage :
T2    1786
T1    1603
T3     533
T4     102
Name: T Stage , dtype: int64

Value counts for N Stage:
N1    2732
N2     820
N3     472
Name: N Stage, dtype: int64

Value counts for 6th Stage:
IIA     1305
IIB     1130
IIIA    1050
IIIC     472
IIIB      67
Name: 6th Stage, dtype: int64

Value counts for differentiate:
Moderately differentiated    2351
Poorly differentiated        1111
Well differentiated           543
Undifferentiated               19
Name: differentiate, dtype: int64

Value counts for Grade:
2                        2351
3                        1111
1                         543
 anaplastic; Grade IV      19
Name: Grade, dtype: int64

Value counts for A Stage:
Regional    3932
Distant       92
Nam

### Gain knowledge about numerical features

In [11]:
numerical_columns = df.select_dtypes(include=['int', 'float']).columns

# Describe numerical columns
numerical_stats = df[numerical_columns].describe()

print("Descriptive statistics for numerical columns:")
print(numerical_stats)

Descriptive statistics for numerical columns:
               Age   Tumor Size  Regional Node Examined  \
count  4024.000000  4024.000000             4024.000000   
mean     53.972167    30.473658               14.357107   
std       8.963134    21.119696                8.099675   
min      30.000000     1.000000                1.000000   
25%      47.000000    16.000000                9.000000   
50%      54.000000    25.000000               14.000000   
75%      61.000000    38.000000               19.000000   
max      69.000000   140.000000               61.000000   

       Reginol Node Positive  Survival Months  
count            4024.000000      4024.000000  
mean                4.158052        71.297962  
std                 5.109331        22.921430  
min                 1.000000         1.000000  
25%                 1.000000        56.000000  
50%                 2.000000        73.000000  
75%                 5.000000        90.000000  
max                46.000000       107

## *
## *
## *
# Label Encoding

In [12]:
df

Age   Race Marital Status T Stage  N Stage 6th Stage  \
0      68  White        Married       T1      N1       IIA   
1      50  White        Married       T2      N2      IIIA   
2      58  White       Divorced       T3      N3      IIIC   
3      58  White        Married       T1      N1       IIA   
4      47  White        Married       T2      N1       IIB   
...   ...    ...            ...      ...     ...       ...   
4019   62  Other        Married       T1      N1       IIA   
4020   56  White       Divorced       T2      N2      IIIA   
4021   68  White        Married       T2      N1       IIB   
4022   58  Black       Divorced       T2      N1       IIB   
4023   46  White        Married       T2      N1       IIB   

                  differentiate Grade   A Stage  Tumor Size Estrogen Status  \
0         Poorly differentiated     3  Regional           4        Positive   
1     Moderately differentiated     2  Regional          35        Positive   
2     Moderately differentiated     2  Regional          63        Positive   
3         Poorly differentiated     3  Regional          18        Positive   
4         Poorly differentiated     3  Regional          41        Positive   
...                         ...   ...       ...         ...             ...   
4019  Moderately differentiated     2  Regional           9        Positive   
4020  Moderately differentiated     2  Regional          46        Positive   
4021  Moderately differentiated     2  Regional          22        Positive   
4022  Moderately differentiated     2  Regional          44        Positive   
4023  Moderately differentiated     2  Regional          30        Positive   

     Progesterone Status  Regional Node Examined  Reginol Node Positive  \
0               Positive                      24                      1   
1               Positive                      14                      5   
2               Positive                      14                      7   
3               Positive                       2                      1   
4               Positive                       3                      1   
...                  ...                     ...                    ...   
4019            Positive                       1                      1   
4020            Positive                      14                      8   
4021            Negative                      11                      3   
4022            Positive                      11                      1   
4023            Positive                       7                      2   

      Survival Months Status  
0                  60  Alive  
1                  62  Alive  
2                  75  Alive  
3                  84  Alive  
4                  50  Alive  
...               ...    ...  
4019               49  Alive  
4020               69  Alive  
4021               69  Alive  
4022               72  Alive  
4023              100  Alive  

[4024 rows x 16 columns]

In [13]:
categorical_columns = ['Status','Marital Status', 'differentiate','Race','T Stage ','Progesterone Status','N Stage','6th Stage','Grade','A Stage','Estrogen Status']  # Your actual feature names
# Encode categorical features
label_encoders = {}
for col in categorical_columns:
    le = LabelEncoder()
    df[col] = le.fit_transform(df[col])
    label_encoders[col] = le
df

Age  Race  Marital Status  T Stage   N Stage  6th Stage  differentiate  \
0      68     2               1         0        0          0              1   
1      50     2               1         1        1          2              0   
2      58     2               0         2        2          4              0   
3      58     2               1         0        0          0              1   
4      47     2               1         1        0          1              1   
...   ...   ...             ...       ...      ...        ...            ...   
4019   62     1               1         0        0          0              0   
4020   56     2               0         1        1          2              0   
4021   68     2               1         1        0          1              0   
4022   58     0               0         1        0          1              0   
4023   46     2               1         1        0          1              0   

      Grade  A Stage  Tumor Size  Estrogen Status  Progesterone Status  \
0         3        1           4                1                    1   
1         2        1          35                1                    1   
2         2        1          63                1                    1   
3         3        1          18                1                    1   
4         3        1          41                1                    1   
...     ...      ...         ...              ...                  ...   
4019      2        1           9                1                    1   
4020      2        1          46                1                    1   
4021      2        1          22                1                    0   
4022      2        1          44                1                    1   
4023      2        1          30                1                    1   

      Regional Node Examined  Reginol Node Positive  Survival Months  Status  
0                         24                      1               60       0  
1                         14                      5               62       0  
2                         14                      7               75       0  
3                          2                      1               84       0  
4                          3                      1               50       0  
...                      ...                    ...              ...     ...  
4019                       1                      1               49       0  
4020                      14                      8               69       0  
4021                      11                      3               69       0  
4022                      11                      1               72       0  
4023                       7                      2              100       0  

[4024 rows x 16 columns]

## *
## *
## *
# Feature selection by Mutual Information

In [14]:
def entropy(labels):
    """Compute entropy of a list of labels."""
    unique, counts = np.unique(labels, return_counts=True)
    probabilities = counts / len(labels)
    return -np.sum(probabilities * np.log2(probabilities))

def mutual_information(X, y):
    """Compute mutual information between feature matrix X and target vector y."""
    n_features = X.shape[1]
    mi_scores = np.zeros(n_features)
    
    # Compute entropy of the target variable y
    H_y = entropy(y)
    
    for i in range(n_features):
        feature_values = X[:, i]
        
        # Calculate entropy of the feature
        H_feature = entropy(feature_values)
        
        # Calculate joint entropy between the feature and target
        joint_entropy = 0
        unique_feature_values = np.unique(feature_values)
        for value in unique_feature_values:
            y_given_feature = y[feature_values == value]
            prob_y_given_feature = len(y_given_feature) / len(y)
            H_y_given_feature = entropy(y_given_feature)
            joint_entropy += prob_y_given_feature * H_y_given_feature
        
        # Calculate mutual information
        mi_scores[i] = H_feature - joint_entropy / H_y
    
    return mi_scores


In [15]:
y=df[['Status']].values
y

array([[0],
       [0],
       [0],
       ...,
       [0],
       [0],
       [0]])

## Selection in numerical features

In [16]:
x_numerical=df[['Age', 'Tumor Size', 'Regional Node Examined','Survival Months','Reginol Node Positive']].values
x_numerical[:5]

array([[68,  4, 24, 60,  1],
       [50, 35, 14, 62,  5],
       [58, 63, 14, 75,  7],
       [58, 18,  2, 84,  1],
       [47, 41,  3, 50,  1]], dtype=int64)

In [17]:
mi_scores = mutual_information(x_numerical, y)
print("Mutual Information Scores:", mi_scores)

Mutual Information Scores: [4.1087501  4.62594051 3.94460649 5.71334976 2.24110172]


### "Regional Node Examined" and "Reginol Node Positive" should be dropped

## Selection in categorical features

In [18]:
x_categorical=df[['Marital Status', 'differentiate','Race','T Stage ','Progesterone Status','N Stage','6th Stage','Grade','A Stage','Estrogen Status']].values
x_categorical[:5]

array([[1, 1, 2, 0, 1, 0, 0, 3, 1, 1],
       [1, 0, 2, 1, 1, 1, 2, 2, 1, 1],
       [0, 0, 2, 2, 1, 2, 4, 2, 1, 1],
       [1, 1, 2, 0, 1, 0, 0, 3, 1, 1],
       [1, 1, 2, 1, 1, 0, 1, 3, 1, 1]])

In [19]:
mi_scores = mutual_information(x_categorical, y)
print("Mutual Information Scores:", mi_scores)

Mutual Information Scores: [ 0.50002274  0.42322869 -0.22649368  0.59750719 -0.30261838  0.27656491
  1.07996462  0.42322869 -0.83436807 -0.61518508]


### "Marital Status" and "6th stage" are selected
## *
## *
## *


# Selecting top features
## There are 3 numerical, 1 categorical, and 1 binary columns

In [20]:
columns_to_drop=['T Stage ','N Stage','differentiate','Race','Grade','A Stage','Estrogen Status','Progesterone Status','Reginol Node Positive','Regional Node Examined']
df.drop(columns=columns_to_drop, inplace=True)  # Use inplace=True to modify the original DataFrame


In [21]:
df

Age  Marital Status  6th Stage  Tumor Size  Survival Months  Status
0      68               1          0           4               60       0
1      50               1          2          35               62       0
2      58               0          4          63               75       0
3      58               1          0          18               84       0
4      47               1          1          41               50       0
...   ...             ...        ...         ...              ...     ...
4019   62               1          0           9               49       0
4020   56               0          2          46               69       0
4021   68               1          1          22               69       0
4022   58               0          1          44               72       0
4023   46               1          1          30              100       0

[4024 rows x 6 columns]

## *
## *
## *
# Categorizing Numerical features

In [22]:
def entropy(y):
    unique_classes, counts = np.unique(y, return_counts=True)
    probabilities = counts / len(y)
    return -np.sum(probabilities * np.log2(probabilities))

# Function to calculate information gain
def information_gain(y, split_point, feature_values):
    left_indices = feature_values <= split_point
    right_indices = feature_values > split_point
    left_entropy = entropy(y[left_indices])
    right_entropy = entropy(y[right_indices])
    p_left = len(y[left_indices]) / len(y)
    p_right = len(y[right_indices]) / len(y)
    return entropy(y) - (p_left * left_entropy + p_right * right_entropy)

# Function to find the best split point for a numerical feature
def find_best_split_point(y, feature_values):
    unique_values = np.sort(np.unique(feature_values))
    best_split = None
    best_gain = -1
    for i in range(len(unique_values) - 1):
        split_point = (unique_values[i] + unique_values[i + 1]) / 2
        gain = information_gain(y, split_point, feature_values)
        if gain > best_gain:
            best_gain = gain
            best_split = split_point
    return best_split, best_gain

# Function to categorize a numerical feature
def categorize_numerical_feature(df, feature_name, target_name):
    feature_values = df[feature_name].values
    target_values = df[target_name].values
    split_point, _ = find_best_split_point(target_values, feature_values)
    df[f'{feature_name}_categorized'] = pd.cut(feature_values, bins=[-np.inf, split_point, np.inf], labels=[f'<= {split_point}', f'> {split_point}'])
    return df


In [23]:
df = categorize_numerical_feature(df, 'Age', 'Status')
df['Age_categorized'].value_counts()

<= 61.5    3027
> 61.5      997
Name: Age_categorized, dtype: int64

In [24]:
df = categorize_numerical_feature(df, 'Tumor Size', 'Status')
df['Tumor Size_categorized'].value_counts()

> 17.5     2848
<= 17.5    1176
Name: Tumor Size_categorized, dtype: int64

In [25]:
df = categorize_numerical_feature(df, 'Survival Months', 'Status')
df['Survival Months_categorized'].value_counts()

> 47.5     3551
<= 47.5     473
Name: Survival Months_categorized, dtype: int64

In [26]:
df

Age  Marital Status  6th Stage  Tumor Size  Survival Months  Status  \
0      68               1          0           4               60       0   
1      50               1          2          35               62       0   
2      58               0          4          63               75       0   
3      58               1          0          18               84       0   
4      47               1          1          41               50       0   
...   ...             ...        ...         ...              ...     ...   
4019   62               1          0           9               49       0   
4020   56               0          2          46               69       0   
4021   68               1          1          22               69       0   
4022   58               0          1          44               72       0   
4023   46               1          1          30              100       0   

     Age_categorized Tumor Size_categorized Survival Months_categorized  
0             > 61.5                <= 17.5                      > 47.5  
1            <= 61.5                 > 17.5                      > 47.5  
2            <= 61.5                 > 17.5                      > 47.5  
3            <= 61.5                 > 17.5                      > 47.5  
4            <= 61.5                 > 17.5                      > 47.5  
...              ...                    ...                         ...  
4019          > 61.5                <= 17.5                      > 47.5  
4020         <= 61.5                 > 17.5                      > 47.5  
4021          > 61.5                 > 17.5                      > 47.5  
4022         <= 61.5                 > 17.5                      > 47.5  
4023         <= 61.5                 > 17.5                      > 47.5  

[4024 rows x 9 columns]

### 0 means "Alive", 1 means "Dead"

## *
## *
## *
# Splitter

In [27]:
df1=df[['Status','Age_categorized','Marital Status','6th Stage','Survival Months_categorized','Tumor Size_categorized']].copy()

In [28]:
df1

Status Age_categorized  Marital Status  6th Stage  \
0          0          > 61.5               1          0   
1          0         <= 61.5               1          2   
2          0         <= 61.5               0          4   
3          0         <= 61.5               1          0   
4          0         <= 61.5               1          1   
...      ...             ...             ...        ...   
4019       0          > 61.5               1          0   
4020       0         <= 61.5               0          2   
4021       0          > 61.5               1          1   
4022       0         <= 61.5               0          1   
4023       0         <= 61.5               1          1   

     Survival Months_categorized Tumor Size_categorized  
0                         > 47.5                <= 17.5  
1                         > 47.5                 > 17.5  
2                         > 47.5                 > 17.5  
3                         > 47.5                 > 17.5  
4                         > 47.5                 > 17.5  
...                          ...                    ...  
4019                      > 47.5                <= 17.5  
4020                      > 47.5                 > 17.5  
4021                      > 47.5                 > 17.5  
4022                      > 47.5                 > 17.5  
4023                      > 47.5                 > 17.5  

[4024 rows x 6 columns]

In [29]:
trainset, testset = train_test_split(df1, test_size=0.3, random_state=3)

In [30]:
print(f'Train set size :{len(trainset)}\nTest set size :{len(testset)}')

Train set size :2816
Test set size :1208


## *
## *
## *
# Building descision tree by ID3

### Functions to Calculate IG

In [31]:
def entropy(y):
    _, counts = np.unique(y, return_counts=True)
    probabilities = counts / len(y)
    return -np.sum(probabilities * np.log2(probabilities))

def information_gain(y, feature_values):
    total_entropy = entropy(y)
    weighted_entropy = 0
    unique_values = np.unique(feature_values)
    
    for value in unique_values:
        subset_indices = feature_values == value
        subset_entropy = entropy(y[subset_indices])
        weighted_entropy += (np.sum(subset_indices) / len(y)) * subset_entropy
        
    return total_entropy - weighted_entropy

### Function to select the best feature to splite

In [32]:
def best_feature_to_split(df, target_name):
    features = df.columns.drop(target_name)
    best_gain = -1
    best_feature = None
    
    for feature in features:
        gain = information_gain(df[target_name], df[feature])
        if gain > best_gain:
            best_gain = gain
            best_feature = feature
            
    return best_feature

### Main function to build the tree

In [33]:
def build_tree(df, target_name):
    # If all target values are the same, return a leaf node
    if len(np.unique(df[target_name])) == 1:
        return df[target_name].iloc[0]
    
    # If there are no more features to split on, return the most common target value
    if len(df.columns) == 1:
        return df[target_name].mode()[0]
    
    # Choose the best feature to split on
    best_feature = best_feature_to_split(df, target_name)
    if best_feature is None:
        return df[target_name].mode()[0]
    
    # Create the tree structure
    tree = {best_feature: {}}
    unique_values = np.unique(df[best_feature])
    
    # Split the dataset and recursively build subtrees
    for value in unique_values:
        subset = df[df[best_feature] == value].drop(columns=[best_feature])
        subtree = build_tree(subset, target_name)
        tree[best_feature][value] = subtree
        
    return tree

In [34]:
tree = build_tree(trainset, 'Status')

In [35]:
tree

{'Survival Months_categorized': {'<= 47.5': {'Age_categorized': {'<= 61.5': {'6th Stage': {0: {'Marital Status': {0: {'Tumor Size_categorized': {'<= 17.5': 0,
          '> 17.5': 1}},
        1: {'Tumor Size_categorized': {'<= 17.5': 0, '> 17.5': 0}},
        2: 0,
        3: {'Tumor Size_categorized': {'<= 17.5': 0, '> 17.5': 1}},
        4: 0}},
      1: {'Marital Status': {0: {'Tumor Size_categorized': {'> 17.5': 0}},
        1: {'Tumor Size_categorized': {'> 17.5': 0}},
        2: 1,
        3: {'Tumor Size_categorized': {'> 17.5': 1}},
        4: {'Tumor Size_categorized': {'> 17.5': 0}}}},
      2: {'Marital Status': {0: {'Tumor Size_categorized': {'> 17.5': 1}},
        1: {'Tumor Size_categorized': {'<= 17.5': 1, '> 17.5': 1}},
        2: 1,
        3: {'Tumor Size_categorized': {'<= 17.5': 0, '> 17.5': 1}},
        4: {'Tumor Size_categorized': {'<= 17.5': 1, '> 17.5': 0}}}},
      3: {'Marital Status': {0: {'Tumor Size_categorized': {'> 17.5': 0}},
        1: 1,
        3: 1}

# Predict

In [36]:
def predict_single(tree, sample):
    while isinstance(tree, dict):
        feature = next(iter(tree))
        value = sample[feature]
        tree = tree[feature].get(value)
        if tree is None:  # Handle cases where the value is not in the tree
            return None  # Or some default value
    return tree

def dataframe_predict(tree, test_set):
    predictions = test_set.apply(lambda x: predict_single(tree, x), axis=1)
    return predictions

In [37]:
sample = {
    'Survival Months_categorized': '> 47.5',
    'Age_categorized': '> 61.5',
    '6th Stage': 0,
    'Tumor Size_categorized': '> 17.5',
    'Marital Status': 0
}
sample2 = {
    'Survival Months_categorized': '<= 47.5',
    'Age_categorized': '> 61.5',
    '6th Stage': 3,
    'Tumor Size_categorized': '> 17.5',
    'Marital Status': 4
}
sample3 = {
    'Survival Months_categorized': '> 47.5',
    'Age_categorized': '> 61.5',
    '6th Stage': 2,
    'Tumor Size_categorized': '> 17.5',
    'Marital Status': 1
}
# Assuming 'tree' is already defined as your decision tree structure
prediction1 = predict_single(tree, sample)
prediction2 = predict_single(tree, sample2)
prediction3 = predict_single(tree, sample3)

print(f"Predicted class for sample 1: {prediction1}\nPredicted class for sample 2: {prediction2}\nPredicted class for sample 3: {prediction3}")

Predicted class for sample 1: 0
Predicted class for sample 2: 1
Predicted class for sample 3: 0


## *
## *
## *
# Evaluation by Accuracy method

In [38]:
def Accuracy(y_pred,y_actual):
    T_np=0
    F_np=0
    for i in range(len(y_pred)):
        if y_pred[i]==y_actual[i]:
            T_np+=1
        else:
            F_np+=1
    acc=T_np/(T_np+F_np)
    return acc

In [39]:
y_testset=testset['Status'].copy()
x_testset=testset.drop(columns=['Status'])
y_testset[:5]

3027    0
1740    0
2878    1
879     0
658     0
Name: Status, dtype: int32

In [40]:
predictions = dataframe_predict(tree, testset)
predictions[:5]

3027    0.0
1740    0.0
2878    0.0
879     0.0
658     0.0
dtype: float64

In [41]:
y_testset=y_testset.tolist()
predictions=predictions.tolist()
acc=Accuracy(predictions,y_testset)
acc

0.8990066225165563

# Accuracy of ID3 algorithm is 90 %

## *
## *
## *
## *
# CART Algorithm Implementation
The CART (Classification and Regression Trees) algorithm is a popular decision tree algorithm used for both classification and regression tasks. Here’s a brief explanation of how the CART algorithm works:

**Basics of CART Algorithm:**

**Splitting Criteria:** CART builds a binary tree by recursively partitioning the data into subsets based on the values of features.

**Objective:** The goal is to minimize impurity or maximize information gain at each split.
Splitting Process:

**Gini Impurity:** Measures the likelihood of an incorrect classification of a randomly chosen element if it was randomly classified according to the distribution of labels in the subset.


**Select Best Split:** CART evaluates all possible splits and selects the one that results in the highest information gain (or lowest impurity) for classification tasks.




## Calculate Gini Impurity

In [42]:
def gini_index(groups, classes):
    # Calculate total number of instances across all groups
    n_instances = float(sum([len(group) for group in groups]))
    gini = 0.0
    for group in groups:
        size = float(len(group))
        if size == 0:
            continue  # Skip empty groups to avoid division by zero
        score = 0.0
        for class_val in classes:
            # Calculate the proportion (probability) of instances in the group that belong to class_val
            p = [row[-1] for row in group].count(class_val) / size
            score += p * p
        # Weighted sum of scores across all classes
        gini += (1.0 - score) * (size / n_instances)
    return gini


### Split the Dataset and Build the Tree

In [43]:
def to_terminal(group):
    """
    Determine the terminal value (predicted output) for a group of data points.
    Returns:
    - Terminal value: Most frequent class label in the group.
    """
    outcomes = [row[-1] for row in group]
    return max(set(outcomes), key=outcomes.count)

def split(node, max_depth, min_size, depth):
    """
    Recursively split nodes of the decision tree based on the best split criteria until stopping conditions are met.
    Returns:
    - None (modifies 'node' dictionary in place).
    """
    left, right = node['groups']
    del(node['groups'])

    # Check if either left or right group is empty
    if not left or not right:
        node['left'] = node['right'] = to_terminal(left + right)
        return

    # Check if maximum depth has been reached
    if depth >= max_depth:
        node['left'], node['right'] = to_terminal(left), to_terminal(right)
        return

    # Process the left child
    if len(left) <= min_size:
        node['left'] = to_terminal(left)
    else:
        node['left'] = get_split(left)
        split(node['left'], max_depth, min_size, depth+1)

    # Process the right child
    if len(right) <= min_size:
        node['right'] = to_terminal(right)
    else:
        node['right'] = get_split(right)
        split(node['right'], max_depth, min_size, depth+1)

def test_split(index, value, dataset):
    """
    Split a dataset into left and right groups based on a split condition.
    Returns:
    - left: List of rows where the feature value at 'index' is less than 'value'.
    - right: List of rows where the feature value at 'index' is greater than or equal to 'value'.
    """
    left, right = list(), list()
    for row in dataset:
        if row[index] < value:
            left.append(row)
        else:
            right.append(row)
    return left, right

def get_split(dataset):
    """
    Identify the best feature and value to split the dataset that minimizes impurity (Gini index).
    Returns:
    - Dictionary with keys:
      - 'index': Index of the feature to split on.
      - 'value': Threshold value for the split condition.
      - 'groups': Tuple containing left and right groups resulting from the split.
    """
    class_values = list(set(row[-1] for row in dataset))
    b_index, b_value, b_score, b_groups = 999, 999, 999, None
    for index in range(len(dataset[0])-1):
        for row in dataset:
            groups = test_split(index, row[index], dataset)
            gini = gini_index(groups, class_values)
            if gini < b_score:
                b_index, b_value, b_score, b_groups = index, row[index], gini, groups
    return {'index': b_index, 'value': b_value, 'groups': b_groups}

def build_tree(train, max_depth, min_size):
    """
    Construct the decision tree recursively from the training dataset.
    Returns:
    - Dictionary representing the root node of the constructed decision tree.
    """
    root = get_split(train)
    split(root, max_depth, min_size, 1)
    return root


### Implement Prediction and Evaluation Functions
The F1 score is a metric used to evaluate the performance of a classification model, especially when dealing with imbalanced classes. It combines both precision and recall into a single metric, providing a balance between these two measures.

In [44]:
def predict(node, row):
    """
    Predict the class label for a given data row using a trained decision tree node.
    Returns:
    - Predicted class label for the data row.
    """
    if row[node['index']] < node['value']:
        if isinstance(node['left'], dict):
            return predict(node['left'], row)
        else:
            return node['left']
    else:
        if isinstance(node['right'], dict):
            return predict(node['right'], row)
        else:
            return node['right']

def evaluate_model(tree, test_set):
    """
    Evaluate a decision tree model on a test dataset and generate predictions.
    Returns:
    - List of predictions for each data point in the test set.
    """
    predictions = list()
    for row in test_set:
        prediction = predict(tree, row)
        predictions.append(prediction)
    return predictions

def accuracy_metric(actual, predicted):
    """
    Calculate the accuracy metric between actual and predicted class labels.
    Returns:
    - Accuracy percentage (float).
    """
    correct = 0
    for i in range(len(actual)):
        if actual[i] == predicted[i]:
            correct += 1
    return correct / float(len(actual)) * 100.0

def f1_score_metric(actual, predicted):
    """
    Calculate the F1 score metric between actual and predicted binary labels.
    Returns:
    - F1 score (float).
    """
    true_positive = sum(1 for i in range(len(actual)) if actual[i] == predicted[i] == 1)
    false_positive = sum(1 for i in range(len(actual)) if predicted[i] == 1 and actual[i] == 0)
    false_negative = sum(1 for i in range(len(actual)) if predicted[i] == 0 and actual[i] == 1)

    if true_positive + false_positive == 0 or true_positive + false_negative == 0:
        return 0.0

    precision = true_positive / (true_positive + false_positive)
    recall = true_positive / (true_positive + false_negative)

    if precision + recall == 0:
        return 0.0

    f1_score = 2 * (precision * recall) / (precision + recall)
    return f1_score

def convert_to_binary_labels(labels):
    """
    Convert threshold strings ('<= 17.5' and '> 17.5') to binary labels (0 and 1).
    Returns:
    - List of binary labels (0 or 1) corresponding to each threshold string.
    """
    binary_labels = []
    for label in labels:
        if label == '<= 17.5':
            binary_labels.append(0)
        elif label == '> 17.5':
            binary_labels.append(1)
        else:
            raise ValueError(f"Unexpected label: {label}")
    return binary_labels


### Run the CART algorithm:

In [45]:
# Split the dataset into training and testing sets
train, test = train_test_split(df1, test_size=0.3, random_state=3)

# Build and evaluate the model
max_depth = 5
min_size = 10
tree = build_tree(train.values.tolist(), max_depth, min_size)

predictions = evaluate_model(tree, test.values.tolist())
actual = [row[-1] for row in test.values.tolist()]

# Convert predictions and actual values to binary labels
predictions_binary = convert_to_binary_labels(predictions)
actual_binary = convert_to_binary_labels(actual)

accuracy = accuracy_metric(actual_binary, predictions_binary)
f1 = f1_score_metric(actual_binary, predictions_binary)

print(f'CART Model Accuracy: {accuracy:.2f}%')
print(f'CART Model F1 Score: {f1:.2f}')

CART Model Accuracy: 86.75%
CART Model F1 Score: 0.91


# Accuracy of CART algorithm is 86.7 %
# F1-Score of CART algorithm is 91%

## *
## *
## *
## *
# C4.5
The C4.5 algorithm is a popular method for generating a decision tree, developed by Ross Quinlan. It's an extension of ID3 algorithm. In this algorithm besides using entropy and information gain, We use gain ratio, Gain ratio adjusts information gain by taking into account the intrinsic information of a split. the formulation is:

$GR(S, A) = \frac{IG(S, A)}{IV(A)}$

where intrinsic value (IV) is:

$
IV(A) = -\sum_{j=1}^{m} \frac{|S_j|}{|S|} \log_2 \left( \frac{|S_j|}{|S|} \right)
$

where:
$
\frac{|S_i|}{|S|} \text{ is the proportion of instances in subset } S_i \text{ compared to the total instances } S.
$

In [46]:
# Splitting the dataset into training (75%) and testing (25%) sets
train_df, test_df = train_test_split(df1, test_size=0.25, random_state=4)

# Define the C4.5 algorithm
def c45_algorithm(df, target_name):
    # If all target values are the same, return a leaf node
    if len(np.unique(df[target_name])) == 1:
        return df[target_name].iloc[0]

    # If there are no more features to split on, return the most common target value
    if len(df.columns) == 1:
        return df[target_name].mode()[0]

    # Choose the best feature to split on using gain ratio
    best_feature = best_feature_to_split_c45(df, target_name)
    if best_feature is None:
        return df[target_name].mode()[0]

    # Create the tree structure
    tree = {best_feature: {}}
    unique_values = np.unique(df[best_feature])

    # Split the dataset and recursively build subtrees
    for value in unique_values:
        subset = df[df[best_feature] == value].drop(columns=[best_feature])
        subtree = c45_algorithm(subset, target_name)
        tree[best_feature][value] = subtree

    return tree

# Define the best feature to split for C4.5
def best_feature_to_split_c45(df, target_name):
    features = df.columns.drop(target_name)
    best_gain_ratio = -1
    best_feature = None

    # Iterate through each feature to calculate gain ratio
    for feature in features:
        gr = gain_ratio(df[target_name], df[feature])
        if gr >= best_gain_ratio:
            best_gain_ratio = gr
            best_feature = feature

    return best_feature

# Define the gain ratio
def gain_ratio(y, feature_values):
    ig = information_gain(y, feature_values)  # Calculate information gain
    iv = intrinsic_value(feature_values)  # Calculate intrinsic value

    # Avoid division by zero by handling zero IV cases
    gain_ratio = ig / iv if iv != 0 else 0

    return gain_ratio

# Calculate the intrinsic value
def intrinsic_value(feature_values):
    unique_values, counts = np.unique(feature_values, return_counts=True)
    probabilities = counts / len(feature_values)
    iv = -np.sum(probabilities * np.log2(probabilities))

    # Replace NaN values with 0
    iv = np.where(np.isnan(iv), 0, iv)

    return iv

# Build the C4.5 tree
c45_tree = c45_algorithm(train_df, 'Status')
c45_tree

{'Survival Months_categorized': {'<= 47.5': {'Age_categorized': {'<= 61.5': {'6th Stage': {0: {'Tumor Size_categorized': {'<= 17.5': {'Marital Status': {0: 1,
          1: 0,
          3: 0}},
        '> 17.5': {'Marital Status': {0: 1, 1: 0, 3: 1, 4: 0}}}},
      1: {'Marital Status': {0: {'Tumor Size_categorized': {'> 17.5': 0}},
        1: {'Tumor Size_categorized': {'> 17.5': 0}},
        2: 1,
        3: {'Tumor Size_categorized': {'> 17.5': 1}},
        4: 0}},
      2: {'Marital Status': {0: {'Tumor Size_categorized': {'> 17.5': 1}},
        1: {'Tumor Size_categorized': {'<= 17.5': 1, '> 17.5': 1}},
        2: 1,
        3: {'Tumor Size_categorized': {'<= 17.5': 0, '> 17.5': 1}},
        4: {'Tumor Size_categorized': {'<= 17.5': 1, '> 17.5': 0}}}},
      3: {'Marital Status': {0: {'Tumor Size_categorized': {'> 17.5': 0}},
        1: 1,
        3: 1}},
      4: {'Marital Status': {0: {'Tumor Size_categorized': {'<= 17.5': 0,
          '> 17.5': 1}},
        1: {'Tumor Size_categ

# Calculate the accuracy of C4.5 algorithm:

In [47]:
# Make predictions
predictions45 = dataframe_predict(c45_tree, test_df)
predictions45 = predictions45.tolist()
y_test_df = test_df['Status'].copy()
y_test_df_list = y_test_df.tolist()

# Calculate accuracy
acc = Accuracy(predictions45, y_test_df_list)
print(f"Accuracy: {acc}")

Accuracy: 0.8836978131212724


# Accuracy of C4.5 algorithm is 88 %